### PR TITLE
Adding initial working lambda to push data to another S3 bucket. 

### PR DESCRIPTION
**Pull request by JoeMoCode**
Filters by specific repos and traffic only. Can take list input in an event to target specific dates. This can be used to backfill data or as the mechanism to trigger the lambda update.

The .json files are examples you will need to manually add to the bucket of the source or something in the destination account. Including all and referencing in the README so it is clear how the code is used.

The .py file included is not directly used. The code is inlined in the Cloudformation script. This seemed like the cleanest way to distribute this as a discrete piece. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by jamesiri — CHANGES_REQUESTED** [review]
Can you please move the content under a directory called 's3_replication' instead of 'custom'? It is less ambiguous and allows for people to see exactly what the feature does without jumping into the content.

**Review by jamesiri — CHANGES_REQUESTED** [review]
Overall, this is a great idea but the implementation and usage need a few tweaks. It also shows that I need to write the guidelines for adding new features so I apologize in advance for not stating this earlier. I still greatly appreciate your contribution!

There are multiple places in the cloudformation template that are hard coded variable for the buckets as well as the fact that there are now two places that the actual lambda code need to be updated if changes are made. This is going to quickly become frustrating and lead us to more problems when trying to add new features. Because of this I have marked the cloudformation portions where variables should be replace the hard coded values. This brings me to my next point.

I would like to see changes around how this feature is used and deployed. Since this isn't something that all users want to use it doesn't make sense to tie it directly into the existing infra code, but it should be ran in a similar manner to ensure consistency in usage. Ideally, this would mean that you have a bash script like [deploy_lambda.sh](https://github.com/amzn/oss-datastore/blob/master/deploy_lambda.sh) that calls supporting scripts in the [infra/bin](https://github.com/amzn/oss-datastore/tree/master/infra/bin) directory which then handles the creation and deployment of this specific set of features. To accomplish this I see the work flow like this:

1. Users would add the information for the buckets they want to duplicate to in the .env file, [here](https://github.com/amzn/oss-datastore/blob/master/sample.env) is the sample that I provide. This would include the ARN info for the replication bucket, the IAM roles, and policies. The root bucket being replicated will always be the one marked in the .env file under S3_ROOT_BUCKET.
I am indifferent on how you want to make the environment variables store this data, I just request that they are grouped in some manner and you can have multiple replications to differing buckets with differing policies.

2. The user will run a shell script from the root of the package, I will probably change this later but since you are kind enough to be the first to contribute just put it in the root for now, that will gather the data from .env and then fill in the variables for the cloudformation template. Adding a sanity check here to verify the bucket names, policy and IAM ARNs are accurate would be super dope and save a ton of headaches. This would also be the place that you merge the index.py into the couldformation template. This will also save a ton of headaches and make maintenance much eaiser.

3. Use the CDK to deploy the cloudformation templates and then have a beer because they now have replication.

I am happy to discuss and show examples that already exist in the package which should make your life a bit easier, feel free to ask me anything you need to get this work done, and thanks again for the contribution!

**Comment by JoeMoCode** [comment]
Sorry for the delay. I finally got to this today. Please have a look! Cheers.

File: infra/s3_replication/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+Before using this. Make sure there is an IAM Role that can interface with the Destination bucket in the destination account. You can see an example in `roleDefinitionDestinationExample.json`. This also requires the AWS CLI installed locally. 
+
+## Run the script:
+
+Upload the pushS3Template.yml in Cloudformation:
+
+1. Ensure the following environment variables are set: `AWS_DEFAULT_REGION`, `AWS_ACCOUNT`, `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_SESSION_TOKEN`.
+2. Set `S3_REPLICATE_OTHER_ROLE_ARN` environment variable to the Role ARN in the destination account that will be assumed by our lambda
+3. Set `S3_REPLICATE_BUCKET_NAME` environment variable to the bucket name in the destination account.
+4. Set `S3_REPLICATE_OTHER_ACCOUNT` environment variable to the destination account ID.
```
[the remark this answers is not in the document]
Comment (JoeMoCode) [reply infra/s3_replication/README.md]: heh, I did this, it just didn't end up in the commit..  Not sure why but I also cannot force add this with git.

File: infra/s3_replication/pushS3Template.yml [hunk 2]
```diff
@@ -14,15 +14,23 @@ Parameters:
   IAMRoleToAssume:
     Type: String
     Description: IAM Role to assume from the other account.
-    Default: arn:aws:iam::328766500510:role/github-replication
-    
+    Default: arn:aws:iam::123456789012:role/github-replication
+  ListOfGithubOrgs:
+    Type: String
+    Description: Comma separated string of Github organizations to use.
+    Default: alexa,alexa-labs,alexa-games
```
[the remark this answers is not in the document]
Comment (JoeMoCode) [reply infra/s3_replication/pushS3Template.yml]: Removed all defaults.

[the remark this answers is not in the document]
Comment (JoeMoCode) [reply infra/s3_replication/pushS3Template.yml]: Also, if they want, this file can be used on its own. Just upload to your account and set the parameters in the GUI

**Comment by JoeMoCode** [comment]
One note with this setup script. It appears not to work on all *nix systems. Thanks for finding this James. The sed command relies on replacing the file contents inline without storing a backup. On some linux systems, -i is not recognized, so it does not work. https://unix.stackexchange.com/questions/13711/differences-between-sed-on-mac-osx-and-other-standard-sed

You can try to drop the '' from the sed commands, but then it won't work on mac.. :)

**Comment by JoeMoCode** [comment]
Added another commit to fix the lambda permission issue on cloudwatch event. Sorry I had forgotten to put in the PR for it! It looks like this is all tacked on to the existing PR.

**Pull request merged**